### PR TITLE
Switch MultiInterfaceController to use variadic templates.

### PR DESCRIPTION
**Pull request by mikepurvis**
As proposed in #298.

C++11 is the supported platform in ROS Kinetic, so this should be a safe change: http://www.ros.org/reps/rep-0003.html#kinetic-kame-may-2016-may-2021

Besides general cleanup, this is of interest to better support composing multiple controllers, especially being able to iterate the interfaces used by the controllers being composed.

### CODE REVIEW

File: controller_interface/include/controller_interface/multi_interface_controller.h [hunk 1]
```diff
@@ -172,25 +172,10 @@ std::string enumerateElements(const T& val,
  * };
  * \endcode
  *
- * \tparam T1 Hardware interface type.
+ * \tparam T... Hardware interface types.
  * This parameter is \e required.
```
Comment (mikepurvis) [inline controller_interface/include/controller_interface/multi_interface_controller.h:146]: I can use a `static_assert` with `sizeof...` to enforce there always being >=1 interface specified, but I don't actually see a hard requirement anywhere for that being the case. Is it truly illegal to have a "controller" which has no interfaces, but still runs some code on the control thread in its update method?

File: controller_interface/include/controller_interface/multi_interface_controller.h [hunk 2]
```diff
@@ -172,25 +172,10 @@ std::string enumerateElements(const T& val,
...
- *
- * \tparam T3 Hardware interface type.
- * This parameter is \e optional. Leave unspecified if controller only claims
- * resources from \e two hardware interfaces.
- *
- * \tparam T4 Hardware interface type.
- * This parameter is \e optional. Leave unspecified if controller only claims
- * resources from \e three hardware interfaces.
- *
- * \pre When specified, template parameters \c T1 to \c T4 must be different
```
Comment (mikepurvis) [inline controller_interface/include/controller_interface/multi_interface_controller.h:190]: This could potentially also be validated using `static_assert`; it wasn't checked previously.

File: controller_manager_tests/CMakeLists.txt [hunk 3]
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 2.8.3)
 project(controller_manager_tests)
 
+set_directory_properties(PROPERTIES COMPILE_OPTIONS "-std=c++11")
```
Comment (mikepurvis) [inline controller_manager_tests/CMakeLists.txt]: For now I've put this here, as it's the only package which actually _requires_ it. Users of controller_interface who don't use `multi_interface_controller.h` don't even need it.

That said, perhaps there should be a separate PR/discussion about doing this across the board.

**Comment by mathias-luedtke** [comment]
I would postpone this to melodic. If I undestood it correctly, c++14 is the default for newer compilers.

File: controller_manager_tests/CMakeLists.txt (hunk 3 above)
Comment (mathias-luedtke) [reply controller_manager_tests/CMakeLists.txt]: But (existing!) users of `multi_interface_controller.h` need to set this (?).

Comment (mikepurvis) [reply controller_manager_tests/CMakeLists.txt]: Yes, that's correct. So perhaps this can just be curiosity for now— we can build our system from a fork until the `melodic-devel` branch is created.

**Comment by mathias-luedtke** [comment]
The header could branch on `BOOST_HAS_VARIADIC_TMPL` (since Boost 1.53), but without the std flags settings. So if you use a newer compiler, you can use the variadic templates. Otherwise the old header gets used. Again, this could be reverted later.

However, this would still be an ABI break, but it does only require a rebuild and no further patches..
@bmagyar: What do you think?

**Comment by mikepurvis** [comment]
I'm not in that much of a rush. We can build from a fork for a few months until there's a `melodic-devel` branch to merge this into.

**Comment by bmagyar** [comment]
Even though we treated ABI breaking pretty liberally in Kinetic, I think it is time to stop that. 

I agree with putting this in a `melodic-devel` branch once it is time for it and pretty happy that you guys can in the mean time build from your fork, it means the new code gets some mileage before it even makes it to our repo! Melodic has been added to `rosdistro` 7 days ago :) 

Feel free to leave this PR open as a reminder!

**Comment by mikepurvis** [comment]
Note that as part of this change, it may actually be possible to make the original `Controller<>` template have this variadic behaviour, and simply alias `MultiInterfaceController` to it (or inherit and deprecate), all without breaking API at all— we would simply add two more init functions signatures like:

```
virtual bool init(T*... hw, ros::NodeHandle&, ros::NodeHandle&);
virtual bool init(T*... hw, ros::NodeHandle&);
```

Then the key is being able to call these correctly from the generic `initRequest`, which I think could be done by recursively packing a `std::tuple` with the correct hardware interface pointers. Will play with it and see what comes up.

But if it worked, the interface to the variadic `MultiInterfaceController<SingleInterface>` would be identical to that of `Controller<SingleInterface>`, with the added bonus of giving multi-interface users a nicer API than having to extract the interfaces themselves from a RobotHW pointer.

**Comment by bmagyar** [comment]
@mikepurvis could you please re-target this PR against `melodic-devel`?

**Comment by mathias-luedtke** [comment]
> could you please re-target this PR against melodic-devel?

@bmagyar: The target branch can be changed easily with the PR edit button (top right).

**Comment by bmagyar** [comment]
Olala, thanks for the heads up! I wasn't aware __I__ could do it!

File: controller_manager_tests/CMakeLists.txt (hunk 3 above)
Comment (mikepurvis) [reply controller_manager_tests/CMakeLists.txt]: Since this is now targeting Melodic, I shall remove the compile options.

**Comment by mikepurvis** [comment]
Would love to see some attention as well to #290, #301, and #302, but I'll let you merge this first and then take care of rebasing those afterward.

**Comment by bmagyar** [comment]
Could you rebase this one please?

**Comment by mikepurvis** [comment]
Done. I don't think the Kinetic CI jobs are going to pass, though— those should be disabled for the `melodic-devel` branch (and the Melodic job is failing due to dependency wait).
